pkg/rbac/rbac.go

```diff
@@ -0,0 +1,287 @@
+package rbac
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"strings"
+
+	"golang.org/x/exp/slices"
+
+	authorizationv1 "k8s.io/api/authorization/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+)
+
+// ACLConfig structure captures the configuration related to acl needed to  perform an action
+// e.g. to view a managedcluster, user need the "get" permission on the  managedcluster resource
+// contains ApiGroup, Resource type but  not Version as rules in ClusterRole do not include/specify version
+type ACLConfig struct {
+	gr   schema.GroupResource
+	verb string
+}
+
+// MetricsACLConfig is an instance of ACLConfig containing config for  Metrics Access
+var MetricsACLConfig = ACLConfig{
+	gr: schema.GroupResource{
+		Group:    "cluster.open-cluster-management.io",
+		Resource: "managedclusters",
+	},
+	verb: "metrics/",
+}
+
+// APIEngine is the rbac api for custom/finer grained RBAC, it should preferably be instantiated through
+// the NewAPIEngine function as it will do any required validation
+// Either or both Config and Client to connect to the Kubernetes cluster need to be configured.
+// any additional configuration needed later on e.g. settings for a cache etc would go in here as well
+
+// Note: Idea is that consumers of this API ( like Observability) will create a single instance of API Engine
+// with the KubeConfig of the Hub Cluster configured and whenever making a rbac api call for a specific user,
+// the  bearerToken of the user is input to the rbac api call. This bearerToken along with Hub cluster config
+// will be used to generate a kubeclient to connect to the cluster and get the rbac info for that user.
+//
+// However, KubeClient (instead of KubeConfig) may also be configured on the Rbac APIEngine,
+// this to allow for the cases where only a  kubeClient is available
+// or when rbac api calls are going to be made only for a single user or serviceaccount throughout.
+
+type APIEngine struct {
+	kubeConfig *rest.Config
+	kubeClient kubernetes.Interface
+}
+
+// kubeConfig  - Kubernetes Cluster configuration
+// kubeClient  - Client to talk to a Kubernetes cluster
+// Either or both Config and Client to connect to the Kubernetes cluster can be set.
+// throws an error when neither is set.
+func NewAPIEngine(kConfig *rest.Config, kClient kubernetes.Interface) (*APIEngine, error) {
```

Comment:
@mprahl -  I am not so convinced about this

At the other location (getKubeClientForUser), because the user tokens may unintentionally overwrite each other, we would need to make a copy

But here, may be the caller changes the settings and wants them to take effect everywhere.  This isn't a service , just a library API, may be fair to expect the changes to take everywhere.  Isn't the point of passing by reference,  to be able to make changes.

Reply:
I think that would be undefined behavior if you change the config after instantiating the struct.

I would like to follow what client-go does for `NewForConfig`:
https://github.com/kubernetes/client-go/blob/babfe96631905312244f34500d2fd1a8a10b186c/kubernetes/clientset.go#L453